recipes/integrate-social-logins.md

```diff
@@ -0,0 +1,581 @@
+Integrate social logins for your API
+====================================
+
+Goals
+-----
+
+Be able to use third-party identity providers (Facebook, Google, GitHub, etc.) to signup and login on our API.
+
+This implementation supports **only** third-parties. The built-in oauth server from Apigility is not usable. That means users **must** use a social account.
+
+This implementation assumes the use of Doctrine (but **not** ZfcUser), but could easily be adapted.
+
+Description
+-----------
+
+Social login is not trivial to setup, because of many constraints and variations of implementation from third-parties. For simplicity sake, we opt for a *web workflow*, as opposed to *client-side workflow*. That means that the end-user will see popups and page redirections.
+
+We also assume that our API will be on its own (sub-)domain, and that the web app making use of the API will be on another (sub-)domain. This add a CORS problem that needs to be dealt with.
+
+The login workflow will be:
+
+1. The page **www**.example.com/login shows a selection of providers
+2. The user select one and a popup opens with an URL to our API (**api**.example.com/hybridauth?provider=Facebook)
+3. The API redirect to the provider login page
+4. After user logged and granted access, the provider redirect to our API (**api**.example.com/hybridauth/callback?provider=Facebook)
+5. The API creates/updates the user in its own database
+6. The API finally create its own token and send it back to the original popup by redirecting to **www**.example.com/receive?token=abcdef
+7. The popup find the token in the URL and forward it to the popup opener via a JavaScript function
+8. Finally the popup close itself, leaving only the original login page which is now in possession of a token that can be used for any subsequent API calls
+9. Optionally, the login page could store the token in local storage, so the login process can be avoided the next time the user visit the site
+
+Implementation of login
+-----------------------
+
+### Server side
+
+First of all install HybridAuth which we'll use to talk to third party providers:
+
+``composer require  hybridauth/hybridauth:dev-3.0.0-Remake``
+
+Then create a new RPC service ``HybridAuth`` via Apigility admin interface. Since we will need two action in the same controller, we'll define the route as ``/hybridauth[/:action]``:
+
+![Create a new RPC service](/asset/apigility-documentation/img/recipes-social-login-rpc-service.png)
+
+In ``HybridAuthController.php``, we create two actions. The first one, ``hybridAuthAction()`` will redirect to the provider. And the second action, ``callbackAction()`` will be used as a callback for the provider. This is where we will create/update user, generate a token, and give it the popup.
+
+```php
+<?php
+
+class HybridAuthController extends AbstractActionController
```

Comment:
Maybe moving lots of business code into a dedicated service/model layer? And inject those into the controller via its factory?